File: NewVisionAI/backend/models/product.py
import enum
from models.base import BaseModel
from config.database import db
from datetime import datetime

class FrameShape(enum.Enum):
    """Enumeration of available frame shapes."""
    RECTANGLE = "rectangle"
    ROUND = "round"
    SQUARE = "square"
    OVAL = "oval"
    CAT_EYE = "cat_eye"
    AVIATOR = "aviator"
    WAYFARER = "wayfarer"
    GEOMETRIC = "geometric"
    OVERSIZED = "oversized"
    RIMLESS = "rimless"

class FrameMaterial(enum.Enum):
    """Enumeration of available frame materials."""
    METAL = "metal"
    PLASTIC = "plastic"
    ACETATE = "acetate"
    TITANIUM = "titanium"
    WOOD = "wood"
    CARBON_FIBER = "carbon_fiber"
    MIXED = "mixed"

class LensType(enum.Enum):
    """Enumeration of available lens types."""
    SINGLE_VISION = "single_vision"
    BIFOCAL = "bifocal"
    PROGRESSIVE = "progressive"
    READING = "reading"
    BLUE_LIGHT = "blue_light"
    SUNGLASSES = "sunglasses"
    POLARIZED = "polarized"
    TRANSITION = "transition"

class Product(BaseModel):
    """Model for eyewear products."""
    
    __tablename__ = 'products'
    
    name = db.Column(db.String(100), nullable=False)
    description = db.Column(db.Text)
    brand = db.Column(db.String(100))
    price = db.Column(db.Float, nullable=False)
    discount_price = db.Column(db.Float)
    sku = db.Column(db.String(50), unique=True, nullable=False)
    stock = db.Column(db.Integer, default=0)
    is_available = db.Column(db.Boolean, default=True)
    frame_shape = db.Column(db.Enum(FrameShape), nullable=False)
    frame_material = db.Column(db.Enum(FrameMaterial), nullable=False)
    frame_color = db.Column(db.String(50))
    lens_type = db.Column(db.Enum(LensType))
    lens_color = db.Column(db.String(50))
    frame_width = db.Column(db.Float)
    temple_length = db.Column(db.Float)
    bridge_width = db.Column(db.Float)
    lens_width = db.Column(db.Float)
    lens_height = db.Column(db.Float)
    weight = db.Column(db.Float)  # in grams
    gender = db.Column(db.String(20))  # 'men', 'women', 'unisex'
    face_shapes = db.Column(db.String(255))  # comma-separated list of suitable face shapes
    thumbnail_url = db.Column(db.String(255))
    image_urls = db.Column(db.Text)  # JSON string of image URLs
    model_3d_url = db.Column(db.String(255))  # URL to the 3D model for virtual try-on
    
    # Relationships
    order_items = db.relationship('OrderItem', back_populates='product')
    
    def __repr__(self):
        return f'<Product {self.name}>'

class OrderStatus(enum.Enum):
    """Enumeration of possible order statuses."""
    PENDING = "pending"
    PROCESSING = "processing"
    SHIPPED = "shipped"
    DELIVERED = "delivered"
    CANCELLED = "cancelled"
    RETURNED = "returned"
    REFUNDED = "refunded"

class Order(BaseModel):
    """Model for customer orders."""
    
    __tablename__ = 'orders'
    
    user_id = db.Column(db.Integer, db.ForeignKey('users.id'), nullable=False)
    order_number = db.Column(db.String(50), unique=True, nullable=False)
    status = db.Column(db.Enum(OrderStatus), default=OrderStatus.PENDING)
    total_amount = db.Column(db.Float, nullable=False)
    discount_amount = db.Column(db.Float, default=0.0)
    shipping_amount = db.Column(db.Float, default=0.0)
    tax_amount = db.Column(db.Float, default=0.0)
    shipping_address = db.Column(db.Text)
    billing_address = db.Column(db.Text)
    payment_method = db.Column(db.String(50))
    payment_id = db.Column(db.String(100))
    shipping_date = db.Column(db.DateTime)
    delivery_date = db.Column(db.DateTime)
    notes = db.Column(db.Text)
    
    # Relationships
    user = db.relationship('User', back_populates='orders')
    items = db.relationship('OrderItem', back_populates='order', cascade='all, delete-orphan')
    
    def __repr__(self):
        return f'<Order {self.order_number}>'

class OrderItem(BaseModel):
    """Model for individual items within an order."""
    
    __tablename__ = 'order_items'
    
    order_id = db.Column(db.Integer, db.ForeignKey('orders.id'), nullable=False)
    product_id = db.Column(db.Integer, db.ForeignKey('products.id'), nullable=False)
    quantity = db.Column(db.Integer, default=1)
    price = db.Column(db.Float, nullable=False)  # Price at time of order
    discount = db.Column(db.Float, default=0.0)
    prescription_data = db.Column(db.Text)  # JSON string of prescription data
    customization_data = db.Column(db.Text)  # JSON string of customization options
    
    # Relationships
    order = db.relationship('Order', back_populates='items')
    product = db.relationship('Product', back_populates='order_items')
    
    def __repr__(self):
        return f'<OrderItem {self.id} for Order {self.order_id}>' 

File: NewVisionAI/iOS/NewVisionAI/Models/Product.swift
import Foundation

// MARK: - Frame Shape
enum FrameShape: String, Codable, CaseIterable {
    case rectangle = "rectangle"
    case round = "round"
    case square = "square"
    case oval = "oval"
    case catEye = "cat_eye"
    case aviator = "aviator"
    case wayfarer = "wayfarer"
    case geometric = "geometric"
    case oversized = "oversized"
    case rimless = "rimless"
    
    var displayName: String {
        switch self {
        case .rectangle: return "Rectangle"
        case .round: return "Round"
        case .square: return "Square"
        case .oval: return "Oval"
        case .catEye: return "Cat Eye"
        case .aviator: return "Aviator"
        case .wayfarer: return "Wayfarer"
        case .geometric: return "Geometric"
        case .oversized: return "Oversized"
        case .rimless: return "Rimless"
        }
    }
}

// MARK: - Frame Material
enum FrameMaterial: String, Codable, CaseIterable {
    case metal = "metal"
    case plastic = "plastic"
    case acetate = "acetate"
    case titanium = "titanium"
    case wood = "wood"
    case carbonFiber = "carbon_fiber"
    case mixed = "mixed"
    
    var displayName: String {
        switch self {
        case .metal: return "Metal"
        case .plastic: return "Plastic"
        case .acetate: return "Acetate"
        case .titanium: return "Titanium"
        case .wood: return "Wood"
        case .carbonFiber: return "Carbon Fiber"
        case .mixed: return "Mixed Materials"
        }
    }
}

// MARK: - Lens Type
enum LensType: String, Codable, CaseIterable {
    case singleVision = "single_vision"
    case bifocal = "bifocal"
    case progressive = "progressive"
    case reading = "reading"
    case blueLight = "blue_light"
    case sunglasses = "sunglasses"
    case polarized = "polarized"
    case transition = "transition"
    
    var displayName: String {
        switch self {
        case .singleVision: return "Single Vision"
        case .bifocal: return "Bifocal"
        case .progressive: return "Progressive"
        case .reading: return "Reading"
        case .blueLight: return "Blue Light"
        case .sunglasses: return "Sunglasses"
        case .polarized: return "Polarized"
        case .transition: return "Transition"
        }
    }
}

// MARK: - Product
struct Product: Codable, Identifiable {
    let id: Int
    let name: String
    let description: String?
    let brand: String?
    let price: Double
    let discountPrice: Double?
    let sku: String
    let stock: Int
    let isAvailable: Bool
    let frameShape: FrameShape
    let frameMaterial: FrameMaterial
    let frameColor: String?
    let lensType: LensType?
    let lensColor: String?
    let frameWidth: Double?
    let templeLength: Double?
    let bridgeWidth: Double?
    let lensWidth: Double?
    let lensHeight: Double?
    let weight: Double?
    let gender: String?
    let faceShapes: String?
    let thumbnailURL: String?
    let imageURLs: String?
    let model3DURL: String?
    
    enum CodingKeys: String, CodingKey {
        case id
        case name
        case description
        case brand
        case price
        case discountPrice = "discount_price"
        case sku
        case stock
        case isAvailable = "is_available"
        case frameShape = "frame_shape"
        case frameMaterial = "frame_material"
        case frameColor = "frame_color"
        case lensType = "lens_type"
        case lensColor = "lens_color"
        case frameWidth = "frame_width"
        case templeLength = "temple_length"
        case bridgeWidth = "bridge_width"
        case lensWidth = "lens_width"
        case lensHeight = "lens_height"
        case weight
        case gender
        case faceShapes = "face_shapes"
        case thumbnailURL = "thumbnail_url"
        case imageURLs = "image_urls"
        case model3DURL = "model_3d_url"
    }
    
    // MARK: - Computed Properties
    
    var actualPrice: Double {
        return discountPrice ?? price
    }
    
    var discountPercentage: Int? {
        guard let discountPrice = discountPrice, discountPrice < price else {
            return nil
        }
        
        let percentage = ((price - discountPrice) / price) * 100
        return Int(percentage)
    }
    
    var isOnSale: Bool {
        return discountPrice != nil && discountPrice! < price
    }
    
    var formattedPrice: String {
        let formatter = NumberFormatter()
        formatter.numberStyle = .currency
        formatter.currencyCode = "USD"
        return formatter.string(from: NSNumber(value: price)) ?? "$\(price)"
    }
    
    var formattedDiscountPrice: String? {
        guard let discountPrice = discountPrice else { return nil }
        
        let formatter = NumberFormatter()
        formatter.numberStyle = .currency
        formatter.currencyCode = "USD"
        return formatter.string(from: NSNumber(value: discountPrice)) ?? "$\(discountPrice)"
    }
    
    var thumbnailImageURL: URL? {
        guard let thumbnailURL = thumbnailURL else { return nil }
        
        if thumbnailURL.hasPrefix("http") {
            return URL(string: thumbnailURL)
        }
        
        return URL(string: "https://api.newvisionai.com/\(thumbnailURL)")
    }
    
    var imageURLList: [URL] {
        guard let imageURLs = imageURLs else { return [] }
        
        do {
            if let jsonArray = try JSONSerialization.jsonObject(with: Data(imageURLs.utf8), options: []) as? [String] {
                return jsonArray.compactMap { urlString in
                    if urlString.hasPrefix("http") {
                        return URL(string: urlString)
                    }
                    return URL(string: "https://api.newvisionai.com/\(urlString)")
                }
            }
        } catch {
            print("Error parsing image URLs: \(error)")
        }
        
        return []
    }
    
    var model3DFileURL: URL? {
        guard let model3DURL = model3DURL else { return nil }
        
        if model3DURL.hasPrefix("http") {
            return URL(string: model3DURL)
        }
        
        return URL(string: "https://api.newvisionai.com/\(model3DURL)")
    }
    
    var suitableFaceShapes: [String] {
        guard let faceShapes = faceShapes else { return [] }
        return faceShapes.split(separator: ",").map { String($0).trimmingCharacters(in: .whitespacesAndNewlines) }
    }
}

// MARK: - ProductSearchFilters
struct ProductSearchFilters {
    var brand: String?
    var frameShape: FrameShape?
    var frameMaterial: FrameMaterial?
    var lensType: LensType?
    var gender: String?
    var minPrice: Double?
    var maxPrice: Double?
    var faceShape: String?
    var sortBy: SortOption = .price
    var sortOrder: SortOrder = .ascending
    
    enum SortOption: String {
        case price = "price"
        case name = "name"
    }
    
    enum SortOrder: String {
        case ascending = "asc"
        case descending = "desc"
    }
    
    var queryParameters: [String: Any] {
        var parameters: [String: Any] = [
            "sort_by": sortBy.rawValue,
            "sort_order": sortOrder.rawValue
        ]
        
        if let brand = brand { parameters["brand"] = brand }
        if let frameShape = frameShape { parameters["frame_shape"] = frameShape.rawValue }
        if let frameMaterial = frameMaterial { parameters["frame_material"] = frameMaterial.rawValue }
        if let lensType = lensType { parameters["lens_type"] = lensType.rawValue }
        if let gender = gender { parameters["gender"] = gender }
        if let minPrice = minPrice { parameters["min_price"] = minPrice }
        if let maxPrice = maxPrice { parameters["max_price"] = maxPrice }
        if let faceShape = faceShape { parameters["face_shape"] = faceShape }
        
        return parameters
    }
}

// MARK: - ProductService
class ProductService {
    static let shared = ProductService()
    
    private init() {}
    
    func getProducts(filters: ProductSearchFilters? = nil, page: Int = 1, completion: @escaping (Result<[Product], Error>) -> Void) {
        var parameters: [String: Any] = ["page": page, "per_page": 20]
        
        if let filters = filters {
            parameters.merge(filters.queryParameters) { (_, new) in new }
        }
        
        APIService.shared.getProducts(parameters: parameters) { result in
            switch result {
            case .success(let json):
                if let productsData = json["products"] as? [[String: Any]] {
                    do {
                        let jsonData = try JSONSerialization.data(withJSONObject: productsData, options: [])
                        let products = try JSONDecoder().decode([Product].self, from: jsonData)
                        completion(.success(products))
                    } catch {
                        completion(.failure(error))
                    }
                } else {
                    completion(.failure(APIError.invalidResponse))
                }
                
            case .failure(let error):
                completion(.failure(error))
            }
        }
    }
    
    func getProduct(id: Int, completion: @escaping (Result<Product, Error>) -> Void) {
        APIService.shared.getProduct(id: id) { result in
            switch result {
            case .success(let json):
                do {
                    let jsonData = try JSONSerialization.data(withJSONObject: json, options: [])
                    let product = try JSONDecoder().decode(Product.self, from: jsonData)
                    completion(.success(product))
                } catch {
                    completion(.failure(error))
                }
                
            case .failure(let error):
                completion(.failure(error))
            }
        }
    }
    
    func getRecommendedProducts(faceShape: String? = nil, frameStyle: String? = nil, color: String? = nil, gender: String? = nil, completion: @escaping (Result<[Product], Error>) -> Void) {
        var parameters: [String: Any] = [:]
        
        if let faceShape = faceShape { parameters["face_shape"] = faceShape }
        if let frameStyle = frameStyle { parameters["frame_style"] = frameStyle }
        if let color = color { parameters["color"] = color }
        if let gender = gender { parameters["gender"] = gender }
        
        APIService.shared.getRecommendedProducts(parameters: parameters) { result in
            switch result {
            case .success(let productsData):
                do {
                    let jsonData = try JSONSerialization.data(withJSONObject: productsData, options: [])
                    let products = try JSONDecoder().decode([Product].self, from: jsonData)
                    completion(.success(products))
                } catch {
                    completion(.failure(error))
                }
                
            case .failure(let error):
                completion(.failure(error))
            }
        }
    }
} 